pgx/src/datum/date.rs

```diff
@@ -8,54 +8,80 @@ Use of this source code is governed by the MIT license that can be found in the
 */
 
 use crate::{pg_sys, FromDatum, IntoDatum};
-use std::ops::{Deref, DerefMut};
-use time::format_description::FormatItem;
+use std::ffi::CStr;
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[repr(transparent)]
+pub struct Date(i32);
+
+impl TryFrom<pg_sys::Datum> for Date {
+    type Error = i32;
+    fn try_from(d: pg_sys::Datum) -> Result<Self, Self::Error> {
+        Ok(Date(d.value() as i32))
+    }
+}
+
+impl IntoDatum for Date {
+    fn into_datum(self) -> Option<pg_sys::Datum> {
+        Some(pg_sys::Datum::from(self.0))
+    }
+    fn type_oid() -> u32 {
+        pg_sys::DATEOID
+    }
+}
 
-#[derive(Debug)]
-pub struct Date(time::Date);
 impl FromDatum for Date {
-    #[inline]
-    unsafe fn from_datum(datum: pg_sys::Datum, is_null: bool) -> Option<Date> {
+    unsafe fn from_datum(datum: pg_sys::Datum, is_null: bool) -> Option<Self>
+    where
+        Self: Sized,
+    {
         if is_null {
             None
         } else {
-            Some(Date(
-                time::Date::from_julian_day(
-                    datum.value() as i32 + pg_sys::POSTGRES_EPOCH_JDATE as i32,
-                )
-                .expect("Unexpected error getting the Julian day in Date::from_datum"),
-            ))
+            Some(datum.try_into().expect("Error converting date datum"))
         }
     }
 }
-impl IntoDatum for Date {
+
+impl Date {
+    pub const NEG_INFINITY: Self = Date(i32::MIN);
+    pub const INFINITY: Self = Date(i32::MAX);
+
     #[inline]
-    fn into_datum(self) -> Option<pg_sys::Datum> {
-        Some((self.to_julian_day() as i32 - pg_sys::POSTGRES_EPOCH_JDATE as i32).into())
+    pub fn from_date(date: time::Date) -> Date {
+        Self::from_pg_epoch_days(date.to_julian_day() - pg_sys::POSTGRES_EPOCH_JDATE as i32)
     }
 
-    fn type_oid() -> u32 {
-        pg_sys::DATEOID
+    #[inline]
+    pub fn from_pg_epoch_days(pg_epoch_days: i32) -> Date {
+        Date(pg_epoch_days)
     }
-}
 
-impl Date {
-    pub fn new(date: time::Date) -> Self {
-        Date(date)
+    #[inline]
+    pub fn is_infinity(&self) -> bool {
+        self == &Self::INFINITY
     }
-}
```

Comment:
We should break Deref and DerefMut, but we shouldn't break `Date::new` immediately. `pub fn new` should be brought back, but with
```rust
#[deprecated(since = "0.5.0",
    note = "the repr of pgx::Date is no longer time::Date \
    and this fn will be removed in a future version")]
```
applied to it.